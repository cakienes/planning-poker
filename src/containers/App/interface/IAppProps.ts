export default interface IAppProps {
    getSessionsFromLocalStorage: () => void;
}
