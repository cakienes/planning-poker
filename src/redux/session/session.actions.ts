import ISession from '../../interfaces/ISession';
import {
    ADD_SESSION,
    GET_SESSIONS_FROM_LOCAL_STORAGE,
    GIVE_STORY_POINT,
    SET_DEVELOPER,
    SET_FINAL_SCORE,
    SET_NEW_DEVELOPER,
    SET_SELECTED_SESSION,
    SET_SELECTED_SESSION_BY_URL,
} from './session.constants';

export const getSessionsFromLocalStorage = () => (dispatch: Function) => {
    dispatch({
        type: GET_SESSIONS_FROM_LOCAL_STORAGE,
    });
};

export const createSession = (session: ISession) => (dispatch: Function) => {
    dispatch({
        type: ADD_SESSION,
        session,
    });
};
export const setSelectedSession = (id: string) => (dispatch: Function) => {
    dispatch({
        type: SET_SELECTED_SESSION,
        id,
    });
};

export const giveStoryPoint = (voterName: string, storyPoint?: number) => (dispatch: Function) => {
    dispatch({
        type: GIVE_STORY_POINT,
        voterName,
        storyPoint,
    });
};

export const setNewDeveloper = (guid: string) => (dispatch: Function) => {
    dispatch({
        type: SET_NEW_DEVELOPER,
        guid,
    });
};
export const setDeveloper = (developerId: string) => (dispatch: Function) => {
    dispatch({
        type: SET_DEVELOPER,
        developerId,
    });
};
export const setFinalScore = (finalScore: string) => (dispatch: Function) => {
    dispatch({
        type: SET_FINAL_SCORE,
        finalScore,
    });
};
export const setSelecedSessionByUrl = (optionalUrl: string) => (dispatch: Function) => {
    dispatch({
        type: SET_SELECTED_SESSION_BY_URL,
        optionalUrl,
    });
};
