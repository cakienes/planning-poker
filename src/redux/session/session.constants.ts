export const GET_SESSIONS_FROM_LOCAL_STORAGE = 'GET_SESSIONS_FROM_LOCAL_STORAGE';
export const ADD_SESSION = 'ADD_SESSION';
export const SET_SELECTED_SESSION = 'SET_SELECTED_SESSION';
export const GIVE_STORY_POINT = 'GIVE_STORY_POINT';
export const SET_NEW_DEVELOPER = 'SET_NEW_DEVELOPER';
export const SET_DEVELOPER = 'SET_DEVELOPER';
