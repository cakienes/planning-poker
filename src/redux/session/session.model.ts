import ISession from '../../interfaces/ISession';

export default interface ISessionModel {
    sessions?: ISession[];
    selectedSessionId?: string;
    developerId?: string;
}
