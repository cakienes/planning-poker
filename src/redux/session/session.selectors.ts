import { createSelector } from 'reselect';
import { UserStoryStatusEnum } from '../../helper/Enum';
import IGlobalState from '../../interfaces/IGlobalState';
import ISession from '../../interfaces/ISession';
import IUserStory from '../../interfaces/IUserStory';
import ISessionModel from './session.model';

export const getState = (state: IGlobalState): IGlobalState => state;
export const getSession = (state: IGlobalState): ISessionModel => state.session;

export const getSessions = createSelector(
    getSession,
    (session: ISessionModel) => {
        return session.sessions;
    },
);

export const getSelectedSessionId = createSelector(
    getSession,
    (session: ISessionModel) => {
        return session.selectedSessionId;
    },
);

export const getSelectedSession = createSelector(
    [getSessions, getSelectedSessionId],
    (sessions: ISession[] | undefined, selectedSessionId: string | undefined) => {
        if (sessions && selectedSessionId) {
            return sessions.find(x => x.id === selectedSessionId);
        }
        return undefined;
    },
);

export const getActiveUserStory = createSelector(
    getSelectedSession,
    (selectedSession?: ISession | undefined) => {
        if (selectedSession && selectedSession.userStories && selectedSession.userStories.length > 0) {
            return selectedSession.userStories.find((x: IUserStory) => x.status === UserStoryStatusEnum.ACTIVE);
        }
    },
);
