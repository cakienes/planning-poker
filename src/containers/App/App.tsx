import React from 'react';
import { connect } from 'react-redux';
import { BrowserRouter, Route } from 'react-router-dom';
import { getSessionsFromLocalStorage } from '../../redux/session/session.actions';
import CreateSession from '../CreateSession/CreateSession';
import SessionError from '../SessionError/SessionError';
import SessionForDeveloper from '../SessionForDeveloper/SessionForDeveloper';
import SessionForScrumMaster from '../SessionForScrumMaster/SessionForScrumMaster';
import './App.scss';
import IAppProps from './interface/IAppProps';

export class App extends React.Component<IAppProps, any> {
    constructor(props: IAppProps) {
        super(props);

        this.props.getSessionsFromLocalStorage();
    }
    render() {
        return (
            <BrowserRouter>
                <Route path="/" exact component={CreateSession} />
                <Route path="/session/:id/developer/:developerId?" exact component={SessionForDeveloper} />
                <Route path="/:optionalUrl/developer/:developerId?" exact component={SessionForDeveloper} />
                <Route path="/session/:id/scrum-master" exact component={SessionForScrumMaster} />
                <Route path="/session-error" exact component={SessionError} />
            </BrowserRouter>
        );
    }
}

export const mapDispatchToProps = (dispatch: Function) => {
    return {
        getSessionsFromLocalStorage: () => {
            dispatch(getSessionsFromLocalStorage());
        },
    };
};

export default connect(
    null,
    mapDispatchToProps,
)(App);
