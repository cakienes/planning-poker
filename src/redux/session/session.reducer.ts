import { UserStoryStatusEnum, UserTypeEnum } from '../../helper/Enum';
import ISession from '../../interfaces/ISession';
import LocalStorageService from '../../services/LocalStorageService';
import * as SessionConstants from './session.constants';
import ISessionModel from './session.model';

export default function SessionReducer(state: ISessionModel = {}, action: any) {
    switch (action.type) {
        case SessionConstants.GET_SESSIONS_FROM_LOCAL_STORAGE: {
            return {
                ...state,
                sessions: LocalStorageService.getSessions(),
            };
        }
        case SessionConstants.ADD_SESSION: {
            const sessions: ISession[] = Object.assign([], state.sessions);
            sessions.push(action.session);
            LocalStorageService.setSessions(sessions);
            return {
                ...state,
                sessions,
            };
        }
        case SessionConstants.SET_SELECTED_SESSION: {
            return {
                ...state,
                selectedSessionId: action.id,
            };
        }
        case SessionConstants.GIVE_STORY_POINT: {
            const sessions: ISession[] = Object.assign([], state.sessions);
            const sessionIndex = sessions.findIndex(x => x.id === state.selectedSessionId);
            if (sessionIndex !== -1) {
                const selectedSession = Object.assign({}, sessions[sessionIndex]);
                if (
                    action.voterName === UserTypeEnum.DEVELOPER &&
                    !selectedSession.developers.find(x => x === state.developerId)
                ) {
                    return state;
                }

                const storyIndex = selectedSession.userStories.findIndex(x => x.status === UserStoryStatusEnum.ACTIVE);

                if (storyIndex !== -1) {
                    const activeStory = Object.assign({}, selectedSession.userStories[storyIndex]);
                    let voterId: string = UserTypeEnum.SCRUM_MASTER;
                    if (action.voterName === UserTypeEnum.DEVELOPER && state.developerId) {
                        voterId = state.developerId;
                    }
                    if (activeStory.voters) {
                        const voterIndex: number = activeStory.voters.findIndex(x => x.voterName === voterId);
                        if (voterIndex === -1) {
                            activeStory.voters.push({
                                storyPoint: action.storyPoint,
                                voterName: voterId,
                            });
                        } else {
                            activeStory.voters[voterIndex] = {
                                storyPoint: action.storyPoint,
                                voterName: voterId,
                            };
                        }
                    } else {
                        activeStory.voters = [
                            {
                                storyPoint: action.storyPoint,
                                voterName: voterId,
                            },
                        ];
                    }

                    selectedSession.userStories[storyIndex] = activeStory;
                }

                sessions[sessionIndex] = selectedSession;
                LocalStorageService.setSessions(sessions);
            }
            return { ...state, sessions };
        }
        case SessionConstants.SET_NEW_DEVELOPER: {
            const sessions: ISession[] = Object.assign([], state.sessions);
            const sessionIndex = sessions.findIndex(x => x.id === state.selectedSessionId);
            sessions[sessionIndex].developers.push(action.guid);
            sessions[sessionIndex] = {
                ...sessions[sessionIndex],
            };

            LocalStorageService.setSessions(sessions);
            return {
                ...state,
                developerId: action.guid,
                sessions,
            };
        }
        case SessionConstants.SET_DEVELOPER: {
            return {
                ...state,
                developerId: action.developerId,
            };
        }
        default: {
            return state;
        }
    }
}
