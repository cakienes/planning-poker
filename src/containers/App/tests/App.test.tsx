import React from 'react';
import { createRenderer, ShallowRenderer } from 'react-test-renderer/shallow';
import { getSessionsFromLocalStorage } from '../../../redux/session/session.actions';
import { App, mapDispatchToProps } from '../App';
import IAppProps from '../interface/IAppProps';

describe('App.tsx', () => {
    let renderer: ShallowRenderer;
    const getSessionsFromLocalStorageMock = jest.fn();
    let appProps: IAppProps = {
        getSessionsFromLocalStorage: getSessionsFromLocalStorageMock,
    };

    beforeEach(() => {
        renderer = createRenderer();
        appProps = {
            getSessionsFromLocalStorage: getSessionsFromLocalStorageMock,
        };
    });
    it('renders correctly UserStoryList', async () => {
        expect(renderer.render(<App {...appProps} />)).toMatchSnapshot();
    });

    it('should call getSessionsFromLocalStorageMock', async () => {
        renderer.render(<App {...appProps} />);
        expect(getSessionsFromLocalStorageMock).toHaveBeenCalled();
    });

    describe('mapDispatchToProps', () => {
        it('should dispatch getLinksFromLocalStorage when called', () => {
            const dispatchMock = jest.fn();
            const result = mapDispatchToProps(dispatchMock);
            result.getSessionsFromLocalStorage();
            expect(JSON.stringify(dispatchMock.mock.calls[0][0])).toEqual(
                JSON.stringify(getSessionsFromLocalStorage()),
            );
        });
    });
});
