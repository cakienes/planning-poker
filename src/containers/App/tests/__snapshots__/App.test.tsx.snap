// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App.tsx renders correctly UserStoryList 1`] = `
<BrowserRouter>
  <Route
    component={[Function]}
    exact={true}
    path="/"
  />
  <Route
    component={[Function]}
    exact={true}
    path="/session/:id/developer/:developerId?"
  />
  <Route
    component={[Function]}
    exact={true}
    path="/:optionalUrl/developer/:developerId?"
  />
  <Route
    component={[Function]}
    exact={true}
    path="/session/:id/scrum-master"
  />
  <Route
    component={[Function]}
    exact={true}
    path="/session-error"
  />
</BrowserRouter>
`;
